float32[] u_t0
float32[] u_tf
float32[] x_t0
float32[] x_tm
float32[] x_tf

float32[] w0_stage1
float32[] w0_stage2
float32[] w0_stage3
float32[] w1_stage1
float32[] w1_stage2
float32[] w1_stage3

float32[] direction
